nixos/modules/services/backup/tsm.nix

```diff
@@ -0,0 +1,358 @@
+
+  assertions = [
+    {
+      assertion = checkIUnique
+        (lib.attrsets.mapAttrsToList (k: v: v.name) cfg.servers);
+      message = ''
+        TSM servernames contain duplicate name
+        (note that case doesn't matter!)
+      '';
+    }
+    {
+      assertion = (cfg.defaultServername!=null)->
+        (lib.attrsets.hasAttr cfg.defaultServername cfg.servers);
+      message = "TSM defaultServername not found in list of servers";
+    }
+    {
+      assertion = (cfgSvc.servername!=null)->
+        (lib.attrsets.hasAttr cfgSvc.servername cfg.servers);
+      message = "TSM service servername not found in list of servers";
+    }
+    {
+      assertion = (cfgSvc.servername!=null)->
+        cfg.servers.${cfgSvc.servername}.genPasswd;
+      message = "TSM service requires automatic password generation";
+    }
+    {
+      assertion = (cfgSvc.autoTime!=null)->(cfgSvc.servername!=null);
+      message = "TSM service timer is set, but its servername is not";
+    }
+  ];
+
+  dsmSysText =
+    let
+      stanzas = lib.attrsets.mapAttrsToList (k: v: v.stanza) cfg.servers;
+    in
+      ''
+        ****  IBM Spectrum Protect (Tivoli Storage Manager)
+        ****  client system-options file "dsm.sys".
+        ****  Do not edit!
+        ****  This file is generated by NixOS configuration.
+
+        ${
+          lib.strings.optionalString (cfg.defaultServername!=null)
+          "defaultserver  ${cfg.defaultServername}"
+        }
+
+        ${lib.strings.concatStringsSep "\n" stanzas}
+      '';
+
+  tsmClientPkg = cfg.package.override rec {
+    dsmSysCli = pkgs.writeText "dsm.sys" cfg.dsmSysText;
+    dsmSysApi = dsmSysCli;
+  };
+
+in
+
+{
+
+  inherit options;
+
+  config = lib.modules.mkIf cfg.enable {
+    inherit assertions;
+    environment.systemPackages = [ tsmClientPkg ];
+    services.tsmClient.dsmSysText = dsmSysText;
+    services.tsmClient.servers = lib.modules.mkIf (cfgSvc.servername!=null) {
+      "${cfgSvc.servername}".passwdDir =
+        lib.modules.mkDefault "/var/lib/tsm-backup/password";
+    };
+    systemd.services.tsm-backup = lib.modules.mkIf (cfgSvc.servername!=null) {
+      description = "IBM Spectrum Protect (Tivoli Storage Manager) Backup";
+      # DSM_LOG needs a trailing slash to have it treated as a directory.
+      # `/var/log` would be littered with TSM log files otherwise.
+      environment.DSM_LOG = "/var/log/tsm-backup/";
+      # TSM needs a HOME dir to store certificates.
+      environment.HOME = "/var/lib/tsm-backup";
+      # for exit status description see
+      # https://www.ibm.com/support/knowledgecenter/en/SSEQVQ_8.1.7/client/c_sched_rtncode.html
+      serviceConfig.SuccessExitStatus = "4 8";
+      # The `-se` option must come after the command.
+      # The `-optfile` option suppresses a `dsm.opt`-not-found warning.
+      serviceConfig.ExecStart =
+        "${tsmClientPkg}/bin/dsmc ${cfgSvc.command} -se='${cfgSvc.servername}' -optfile=/dev/null";
```

Comment:
Is the `dsm.opt` not-found guard still necessary?

Reply:
Yes.  It merely hides a warning, but a long and confusing one.